kernels/portable/cpu/util/test/broadcast_indexes_range_test.cpp

```diff
@@ -0,0 +1,174 @@
+ */
+
+#include <executorch/kernels/portable/cpu/util/broadcast_indexes_range.h>
+#include <executorch/kernels/portable/cpu/util/broadcast_util.h>
+#include <executorch/runtime/core/exec_aten/testing_util/tensor_factory.h>
+
+#include <gtest/gtest.h>
+
+using executorch::aten::ScalarType;
+using executorch::aten::Tensor;
+using executorch::runtime::testing::TensorFactory;
+using torch::executor::BroadcastIndexesRange;
+using torch::executor::delinearize_index;
+using torch::executor::linearize_access_indexes;
+
+namespace {
+template <typename Range>
+auto range_to_vec(const Range& rng) {
+  return std::vector<typename Range::iterator::value_type>(
+      rng.begin(), rng.end());
+}
+} // namespace
+TEST(BroadcastIndexesRangeTest, Empty) {
+  TensorFactory<ScalarType::Int> tf;
+
+  Tensor a = tf.make({0}, {});
+  ASSERT_EQ(a.numel(), 0);
+  bool loop_entered = false;
+  for (auto _ : BroadcastIndexesRange<1>(a, a)) {
+    loop_entered = true;
+  }
+  EXPECT_FALSE(loop_entered);
+}
+
+// [W] -> [W]
+TEST(BroadcastIndexesRangeTest, OneDNotBroadcasted) {
+  TensorFactory<ScalarType::Int> tf;
+
+  Tensor out = tf.zeros({5});
+  int idx = 0;
+  for (const auto& elem : range_to_vec(BroadcastIndexesRange<1>(out, out))) {
+    EXPECT_EQ(elem[0], idx++);
+    EXPECT_EQ(elem[0], elem[1]);
+  }
+}
+
+// [1] -> [W]
+TEST(BroadcastIndexesRangeTest, ScalarBroadcastToOneD) {
+  TensorFactory<ScalarType::Int> tf;
+
+  Tensor out = tf.zeros({5});
+  Tensor in = tf.zeros({1});
+
+  auto actual = range_to_vec(BroadcastIndexesRange<1>(out, in));
+  decltype(actual) expected = {
+      {0, 0},
+      {1, 0},
+      {2, 0},
+      {3, 0},
+      {4, 0},
+  };
+  EXPECT_EQ(expected, actual);
+}
+
+// [1] -> [H, W]
+// [1, W] -> [H, W]
+// [H, 1] -> [H, W]
+// [H, W] -> [H, W]
+// Cover all these at the same time to also exercise multiple input tensors.
+TEST(BroadcastIndexesRangeTest, TwoDExhaustive) {
+  TensorFactory<ScalarType::Int> tf;
+  Tensor out = tf.zeros({3, 4});
+  Tensor in_0d_scalar = tf.zeros({});
+  Tensor in_1d_scalar = tf.zeros({1});
+  Tensor in_2d_scalar = tf.zeros({1, 1});
+
+  Tensor in_row = tf.zeros({4});
+  Tensor in_col = tf.zeros({3, 1});
+
+  Tensor in_not_broadcast = tf.zeros({3, 4});
+
+  auto actual = range_to_vec(BroadcastIndexesRange<6>(
+      out,
+      in_0d_scalar,
+      in_1d_scalar,
+      in_2d_scalar,
+      in_row,
+      in_col,
+      in_not_broadcast));
+  decltype(actual) expected = {
+      {0, 0, 0, 0, 0, 0, 0},
+      {1, 0, 0, 0, 1, 0, 1},
+      {2, 0, 0, 0, 2, 0, 2},
+      {3, 0, 0, 0, 3, 0, 3},
+      {4, 0, 0, 0, 0, 1, 4},
+      {5, 0, 0, 0, 1, 1, 5},
+      {6, 0, 0, 0, 2, 1, 6},
+      {7, 0, 0, 0, 3, 1, 7},
+      {8, 0, 0, 0, 0, 2, 8},
+      {9, 0, 0, 0, 1, 2, 9},
+      {10, 0, 0, 0, 2, 2, 10},
+      {11, 0, 0, 0, 3, 2, 11},
+  };
+  EXPECT_EQ(expected, actual);
+}
+
+// Here we assume that the previous tests established that padding
+// with leading 1s is working, and test:
```

Comment:
You are testing 5 out of 8 possibilities here. You might as well add the remaining 3:
[1, 1, 1] -> [C, H, W]
[1, 1, W] -> [C, H, W] (this one in particular would be good to have, i.e. multiple leading ones)
[1, H, W] -> [C, H, W]